### PR TITLE
As/fix eip155

### PR DESCRIPTION
**Pull request by alsala**

### CODE REVIEW

File: sdk/src/main/java/com/horizen/account/transaction/AccountTransaction.java [hunk 1]
```diff
@@ -24,5 +24,7 @@
 
     public abstract byte[] getData();
 
+    public abstract PR getRealSignature();
```
Comment (i-Alex) [inline sdk/src/main/java/com/horizen/account/transaction/AccountTransaction.java:27]: It is a bit strange to see both `getRealSignature` and `getSignature` methods in the `AccountTransaction`. Let's have only `getSignature()`.

In EthTransaction we may rename `getRealSignature` to `getSignature`. And don't use the current `getSignature` at all. Instead we may define a method like `getSignatureData` that will return `Sign.SignatureData` and used in all the places needed: `toString`/RPC/serialization

File: sdk/src/main/java/com/horizen/account/transaction/EthereumTransaction.java [hunk 2]
```diff
@@ -221,7 +221,7 @@ else if (this.isSigned()) {
             var sigData = signedTx.getSignatureData();
             if (sigData.getS()[0] == 0 && sigData.getR()[0] == 0)
                 return convertToLong(sigData.getV());
-            else return ((SignedRawTransaction) this.transaction).getChainId(); // TODO check necessity
+            else return ((SignedRawTransaction) this.transaction).getChainId(); // useful for EIP155
```
Comment (i-Alex) [inline sdk/src/main/java/com/horizen/account/transaction/EthereumTransaction.java:224]: do we really need if/else here? Can't we use only `((SignedRawTransaction) this.transaction).getChainId()` ?

File: sdk/src/main/java/com/horizen/account/utils/Secp256k1.java [hunk 3]
```diff
@@ -5,9 +5,10 @@ public final class Secp256k1 {
     public static final int PRIVATE_KEY_SIZE = COORD_SIZE;
     public static final int PUBLIC_KEY_SIZE = COORD_SIZE * 2;
 
-    public static final int SIGNATURE_V_SIZE = 1;
+    // signature's v might be set to {0,1} + CHAIN_ID * 2 + 35 (see EIP155)
+    public static final int SIGNATURE_V_MAXSIZE = 4;
     public static final int SIGNATURE_RS_SIZE = 32;
-    public static final int SIGNATURE_SIZE = SIGNATURE_V_SIZE + SIGNATURE_RS_SIZE * 2;
+    public static final int SIGNATURE_MAXSIZE = SIGNATURE_V_MAXSIZE + SIGNATURE_RS_SIZE * 2;
```
Comment (i-Alex) [inline sdk/src/main/java/com/horizen/account/utils/Secp256k1.java:11]: actually never used. Can we removed it?

File: sdk/src/test/java/com/horizen/account/transaction/EthereumTransactionTest.java [hunk 4]
```diff
@@ -196,6 +211,27 @@ public void ethereumLegacyEIP155TransactionTest() {
...
+        byte[] toAddress = decodedTx.getTo().address();
+
+        assertEquals("892278d9f50a1da5b2e98e5056f165b1b2486d97", BytesUtils.toHexString(fromAddress));
+        assertEquals("d830603264bd3118cf95f1fc623749337342f9e9", BytesUtils.toHexString(toAddress));
+        assertEquals(1997, chainId);
+        assertEquals("3000000000000000000", decodedTx.getValue().toString());
+
+        // re-encode and check it is the same
+        byte[] encodedTx = EthereumTransactionSerializer.getSerializer().toBytes(decodedTx);
+        assertEquals(metamaskHexStr, BytesUtils.toHexString(encodedTx));
```
Comment (i-Alex) [inline sdk/src/test/java/com/horizen/account/transaction/EthereumTransactionTest.java:234]: let's check the signature before and after encoding

**Review by i-Alex — CHANGES_REQUESTED** [review]

File: sdk/src/main/java/com/horizen/account/transaction/AccountTransaction.java (hunk 1 above)
Comment (alsala) [reply sdk/src/main/java/com/horizen/account/transaction/AccountTransaction.java]: Ok, done

File: sdk/src/main/java/com/horizen/account/transaction/EthereumTransaction.java (hunk 2 above)
Comment (alsala) [reply sdk/src/main/java/com/horizen/account/transaction/EthereumTransaction.java]: I added some comment

File: sdk/src/main/java/com/horizen/account/utils/Secp256k1.java (hunk 3 above)
Comment (alsala) [reply sdk/src/main/java/com/horizen/account/utils/Secp256k1.java]: Right, done

File: sdk/src/test/java/com/horizen/account/transaction/EthereumTransactionTest.java (hunk 4 above)
Comment (alsala) [reply sdk/src/test/java/com/horizen/account/transaction/EthereumTransactionTest.java:237]: Ok, done

File: sdk/src/main/java/com/horizen/account/transaction/EthereumTransaction.java [hunk 5]
```diff
@@ -310,18 +304,21 @@ public SignatureSecp256k1 getRealSignature() {
...
                     stx.getSignatureData().getV(),
                     stx.getSignatureData().getR(),
                     stx.getSignatureData().getS());
         }
         return null;
     }
 
+    public byte[] getV() { return (getSignatureData() != null) ? getSignatureData().getV() : null; }
+    public byte[] getR() { return (getSignatureData() != null) ? getSignatureData().getR() : null; }
+    public byte[] getS() { return (getSignatureData() != null) ? getSignatureData().getS() : null; }
```
Comment (i-Alex) [inline sdk/src/main/java/com/horizen/account/transaction/EthereumTransaction.java:321]: I think it is never used.
let's remove this methods, of?

File: sdk/src/main/java/com/horizen/account/transaction/EthereumTransaction.java [hunk 6]
```diff
@@ -349,22 +346,30 @@ public String toString() {
...
+        ByteBuffer buffer = ByteBuffer.allocate(Long.BYTES);
+        buffer.putLong(x);
+        return buffer.array();
+    }
+
+    // w3j way for converting bytes, it works also with generic byte contents (not only Long.BYTES byte arrays)
+    public static long convertToLong(byte[] bytes) {
+        BigInteger bi = Numeric.toBigInt(bytes);
+        return bi.longValue();
+    }
```
Comment (i-Alex) [inline sdk/src/main/java/com/horizen/account/transaction/EthereumTransaction.java:374]: I would suggest to move it a separate utility class into `com.horizen.account.utils` package.

**Review by i-Alex — CHANGES_REQUESTED** [review]

File: sdk/src/main/java/com/horizen/account/transaction/EthereumTransaction.java (hunk 5 above)
Comment (alsala) [reply sdk/src/main/java/com/horizen/account/transaction/EthereumTransaction.java:322]: They are used in json representation, I added a comment

File: sdk/src/main/java/com/horizen/account/transaction/EthereumTransaction.java (hunk 6 above)
Comment (alsala) [reply sdk/src/main/java/com/horizen/account/transaction/EthereumTransaction.java]: Ok, moved in a new EthereumTransactionUtils class in the package

**Review by JcBernack — APPROVED** [review]

File: sdk/src/main/scala/com/horizen/account/api/http/AccountTransactionApiRoute.scala [hunk 7]
```diff
@@ -124,16 +137,33 @@ case class AccountTransactionApiRoute(override val settings: RESTApiSettings,
...
+            val response = if (isEIP155) {
+                val tmpTx = new EthereumTransaction(
+                  destAddress,
+                  nonce,
+                  gasPrice,
+                  gasLimit,
+                  valueInWei,
+                  "",
+                  new SignatureData(
+                    EthereumTransaction.convertToBytes(params.chainId),
```
Comment (JcBernack) [inline sdk/src/main/scala/com/horizen/account/api/http/AccountTransactionApiRoute.scala:151]: moved to `EthereumTransactionUtils`

**Review by JcBernack — CHANGES_REQUESTED** [review]

File: sdk/src/main/scala/com/horizen/account/api/http/AccountTransactionApiRoute.scala (hunk 7 above)
Comment (alsala) [reply sdk/src/main/scala/com/horizen/account/api/http/AccountTransactionApiRoute.scala]: Added, thanks!

**Review by i-Alex — APPROVED** [review]

**Review by JcBernack — APPROVED** [review]

**Pull request merged**